# mapping

Here is an example config.json:

```json
{
    "base": "/lustre_cfc/ws_qstor/ws/afaaj01-QLBAB_EXOME-0",
    "data": "/lustre_cfc/ws_qstor/ws/afaaj01-QLBAB_EXOME-0/data",
    "ref": "/lustre_cfc/ws_qstor/ws/afaaj01-QLBAB_EXOME-0/ref",
    "src": "/lustre_cfc/ws_qstor/ws/afaaj01-QLBAB_EXOME-0/src",
    "var": "/lustre_cfc/ws_qstor/ws/afaaj01-QLBAB_EXOME-0/var",
    "result": "/lustre_cfc/ws_qstor/ws/afaaj01-QLBAB_EXOME-0/result",
    "run": "/lustre_cfc/ws_qstor/ws/afaaj01-QLBAB_EXOME-0/run",
    "etc": "/lustre_cfc/ws_qstor/ws/afaaj01-QLBAB_EXOME-0/etc",
    "logs": "/lustre_cfc/ws_qstor/ws/afaaj01-QLBAB_EXOME-0/logs",
    "archive": "/lustre_cfc/ws_qstor/ws/afaaj01-QLBAB_EXOME-0/archive",
    "usr": "/lustre_cfc/ws_qstor/ws/afaaj01-QLBAB_EXOME-0/usr",
    "params": {"fasta": "/lustre_cfc/qbic/reference_genomes/Homo_sapiens/DNA/2016.01.21.UCSC/hg19/Sequence/BWAIndex/hg19/hg19"} 
}
```

Note that in contrast to the RNA-Seq workflow you do not need a *params.json* file here.
